### ISSUE TITLE
Dynamic link error with Android less then 29 API level

### ISSUE THREAD

**Issue by Gordon-F** [timothee-haudebourg/khronos-egl#11]
Test [gfx-quad gl](https://github.com/gfx-rs/gfx) example on Android:

```
2020-12-21 22:38:15.970 9618-9727/rust.quad_android I/RustStdoutStderr: NativeScreen Found:RwLockReadGuard { lock: RwLock { data: Some(NativeWindow { ptr: 0x7d3f1fd010 }) } }
2020-12-21 22:38:15.972 9618-9728/rust.quad_android W/gfx_backend_gl::windo..: Unable to open libEGL.so: DlSym { desc: "undefined symbol: eglGetPlatformDisplay" }
```

eglGetPlatformDisplay only avaliable in API level 29 (Android 10).
https://developer.android.google.cn/reference/android/opengl/EGL15#eglGetPlatformDisplay(int,%20long,%20long[],%20int)

Maybe we can feature-gate egl version api at compile time?

**Comment by timothee-haudebourg** [comment]
Depends... do you always know at compile time what version you are running on?

**Comment by timothee-haudebourg** [comment]
With the `versions` branch you should be able to choose a specific API version at compile time. Can you try it out?

For instance in your case, if you only want to use EGL 1.4, use something like this in your `Cargo.toml`:
```toml
[dependencies.khronos-egl]
git = "https://github.com/timothee-haudebourg/khronos-egl.git"
branch = "versions"
default-features = false
features = ["dynamic", "1.4"] # dynamic loading, up to EGL 1.4
```
then you should be able to use `DynamicInstance` as usual.

Alternatively, you can load a specific version of EGL using the type system:
```rust
use khronos_egl::{Dynamic, Instance, EGL1_4};

let lib = libloading::Library::new(filename)?;
let egl = Instance::new(Dynamic::<_, EGL1_4>::new(lib)?); // only EGL1.4 will be loaded here.
```

You can also write functions for a specific API version, although it may not be that useful until function specialization becomes a thing in Rust:
```rust
use khronos_egl::{Instance, api};

fn foo<V: api::EGL1_4>(egl: &Instance<V>) {
  // do something that requires at least EGL 1.4.
}
```

**Comment by Gordon-F** [comment]
@timothee-haudebourg thanks! I think we can't always know version at compile time.

> I may try to add something to load the most recent available version dynamically.

That would be awesome! We can check available version on client side and use only available api.

**Comment by timothee-haudebourg** [comment]
Here you go. You can now load `libEGL.so` with minimal requirements, and then cast between different versions whenever you need.

Here is an example that I have added:
```rust
extern crate gl;
extern crate khronos_egl as egl;
extern crate libloading;

use std::sync::Arc;
use egl::{
	EGL1_2,
	EGL1_4
};

fn main() {
	let minimal_egl = unsafe { Arc::new(egl::DynamicInstance::load_minimal().expect("unable to load libEGL.so")) };

	println!("EGL version is {}", minimal_egl.version());

	// Select the rendering API.
	if let Some(egl1_2) = minimal_egl.upcast::<EGL1_2>() {
		 egl1_2.bind_api(egl::OPENGL_API).expect("unable to select OpenGL API");
	}

	// Setup Open GL.
	gl::load_with(|name| minimal_egl.get_proc_address(name).unwrap() as *const std::ffi::c_void);

	match minimal_egl.upcast::<EGL1_4>() {
		Some(egl1_4) => {
			foo_with_1_4(egl1_4)
		},
		None => {
			foo_without_1_4(&minimal_egl)
		}
	}
}

fn foo_with_1_4<V: egl::api::EGL1_4>(_egl: &egl::Instance<V>) {
	// do something that requires at least EGL 1.4.
}

fn foo_without_1_4<V>(_egl: &egl::Instance<V>) {
	// do something without any specific EGL version (other that 1.0).
}
```

The `DynamicInstance::load_minimal` function will load at least EGL 1.0. You can then upgrade it to the version that fits your needs dynamically using `DynamicInstance::upcast`/`DynamicInstance::downcast` (by reference) or `DynamicInstance::cast`/`DynamicInstance::try_cast` (which will consume the instance).
The most recent available version can be queried using `DynamicInstance::version`. In your case it should show `1.4`.

**Comment by timothee-haudebourg** [comment]
@Gordon-F does that suit your needs? Tell me if it works for you or not so I can release version 3.0

**Comment by Gordon-F** [comment]
@kvark Do you have any suggestions about new api?

**Comment by kvark** [comment]
Effectively for gfx-rs this API would mean that we have
```rust
enum Egl {
  V14(egl::Instance<egl::api::EGL1_4>), 
  V15(egl::Instance<egl::api::EGL1_5>), 
}
```
And then we'd have to write a bunch of boilerplate to call methods on it... Definitely less convenient than we'd like. I.e. there is no benefit from this API to gfx-rs, it would only stand in the way.

Could we have a way to make it less type-driven? For example,
```rust
impl DynamicInstance {
  fn load() -> (Self, Version);
```
I.e. this will try to load the highest version entry points and give you the version together with the library. That would be most convenient to us.

**Comment by timothee-haudebourg** [comment]
> I.e. this will try to load the highest version entry points and give you the version together with the library. That would be most convenient to us.

This is already what `DynamicInstance::load_minimal` is doing (maybe the name isn't really clear). It will load the the highest available version entry points. The returned type is `DynamicInstance<EGL1_0>` which means that it provides **at least** EGL 1.0.  You can query what version is actually loaded using `DynamicInstance::version()`. It can be greater than the version given by the type, in which case you can upgrade it to a `DynamicInstance<EGL1_X>` using the casting functions mentioned above.

I especially designed it so you should need such enum type. Is that clearer or did I misunderstood your point?

**Comment by kvark** [comment]
> maybe the name isn't really clear

It's actually confusing, since the logic we are describing is that it tries to get the *maximum* supported version, but it's called "minimum" :)

> It can be greater than the version given by the type, in which case you can upgrade it to a DynamicInstance<EGL1_X> using the casting functions mentioned above.

We would like to avoid that cast, if possible. Working with an enum at this level is just inconvenient. EGL has a thousand ways to shoot yourself in the foot, calling a method that's not available on your version is one of many. Does it worth making the API harder to use then, given that it's unsafe anyway?

At the end of the day, you can check what Ash is doing and follow. We can manage either way you approach this.

**Comment by timothee-haudebourg** [comment]
> It's actually confusing, since the logic we are describing is that it tries to get the maximum supported version, but it's called "minimum" :)

Yeah I agree, coming back to it days later, the name is weird. It think I'll change `load_minimum` to just `load` (and rename the current `load` into `load_exact`).

> We would like to avoid that cast, if possible

I'm not sure to see how to avoid it. If you don't know statically what version is available, you have to dynamically cast what you get.

> EGL has a thousand ways to shoot yourself in the foot, calling a method that's not available on your version is one of many.

Well, that's why casting functions are necessary, so you cannot call methods that are not available.

To be clear, `DynamicInstance<EGL1_X>` gives you access to all the EGL functions **up to** EGL 1.X, but it also store what is the most recent available version, which can be greater that 1.X, in which case you can upcast it.

If you require EGL 1.4 throughout the entire application, then you only have to cast once to `DynamicInstance<EGL1_4>`:
```rust
let any_egl = unsafe { Arc::new(egl::DynamicInstance::load().expect("unable to load libEGL.so")) };

println!("EGL version is {}", any_egl.version());

let egl = match any_egl.upcast::<EGL1_4>() {
	Some(egl) => egl,
	None => panic!("EGL 1.4 or greater is required")
};

// continue with `egl` providing EGL 1.4 as before.
```

**Comment by timothee-haudebourg** [comment]
I should add that casting operations are free.

**Comment by kvark** [comment]
We are going to detect the version and use 1.5 if available. That means calling any 1.4 methods would have to be done with a match over an enum containing at least 1.4 and 1.5...

> If you don't know statically what version is available, you have to dynamically cast what you get.

I thought your function pointers could just be NULL, in which case calling a method panics. That way you'd be able to just allow us to call any methods and detect the version dynamically ourselves.

**Comment by timothee-haudebourg** [comment]
> That means calling any 1.4 methods would have to be done with a match over an enum containing at least 1.4 and 1.5...

Sorry I don't understand why you think you need an enum here. Once it is casted to `DynamicInstance<EGL1_5>` you can call any EGL function, including functions of previous API versions.

```rust
let egl = match any_egl.upcast::<EGL1_5>() {
	Some(egl) => egl,
	None => panic!("EGL 1.5 or greater is required")
};

// you can call any EGL function on `egl`, including EGL 1.0, 1.1, 1.2, 1.3, 1.4 and 1.5 functions.
```

> I thought your function pointers could just be NULL, in which case calling a method panics. That way you'd be able to just allow us to call any methods and detect the version dynamically ourselves.

Internally yes, `DynamicInstance` will hold some null pointers if the latest EGL version is not available. That's why I use the type system (with `<EGL1_X>`) to forbid access to those functions when they are not available.

**Comment by kvark** [comment]
> Sorry I don't understand why you think you need an enum here. Once it is casted to DynamicInstance<EGL1_5> you can call any EGL function, including functions of previous API versions.

So somewhere along the line I wouldn't know at compile time what the version is. So I'd be having an enum like
```rust
enum Egl {
  V14(egl::Instance<egl::api::EGL1_4>), 
  V15(egl::Instance<egl::api::EGL1_5>), 
}
```

And then whenever I need to call anything, I'd have to match against this enum. This is less convenient, obviously, then just being able to call methods on a thing (that would panic if they are `null` pointers).

**Comment by timothee-haudebourg** [comment]
I think you don't understand what I'm providing here. Just to be clear: you do not need such an enum type. `egl::DynamicInstance::load()` (currently named `load_minimal`) will load the most recent available EGL API version, dynamically.

If I understand your settings, you require at least EGL 1.4, but sometime would like to call EGL 1.5 whenever possible. You can do that using `DynamicInstance::try_cast` (to consume) or `DynamicInstance::upcast` (to borrow).

Once loaded, you can directly try to upcast it to the minimum version required by your application:
```rust
let egl: DynamicInstance<EGL1_4> = any_egl.try_cast().expect("EGL 1.4 or greater is required");
```

Here you can call any EGL 1.4 function on `egl` **without any risk of panicking**, since the upcast checks that EGL 1.4 is provided by the instance.
Then, locally, if you want to use an EGL 1.5 whenever available, then you can try to upcast it again:
```rust
match egl.upcast::<EGL1_5>() {
	Some(egl1_5) => // do something with EGL 1.5
	None => // do something with EGL 1.4 instead.
};
```
The type system ensures that you will never be able to call a null pointer. And `DynamicInstance` holds all the information you need on the loaded EGL version, you don't need an enum.

**Comment by kvark** [comment]
So you are suggesting, basically, that the way I store an instance is only by the minimally supported version (e.g. `DynamicInstance<EGL1_4>`), and then upcast whenever I need anything extra? That would work for me indeed.

**Comment by timothee-haudebourg** [comment]
Yes that's the idea :smile: 

Use `DynamicInstance<EGL1_X>` where 1.X is the minimum required version for your app. You will still be able to upcast it whenever you need anything extra, without additional cost.

**Comment by timothee-haudebourg** [comment]
At least this conversation shows that I need to work harder on the documentation :smile:

**Comment by niklaskorz** [comment]
I have updated the gfx-backend-gl code to use `load_minimal` and upcasting, but I am still receiving the following error from `load_minimal` in https://github.com/gfx-rs/gfx/pull/3591/files#diff-3c3708b8ab2a59f5c903de589330d635515c80df4ccc5071e0d2db70a1e77227R226:

```
Unable to open libEGL.so: DlSym { desc: "undefined symbol: eglCreateSync" }
```

**Comment by niklaskorz** [comment]
I have found the issue. Both DlSymUnknown and DlSym need to be catched, I will open a pull request in a moment.

**Comment by timothee-haudebourg** [comment]
In my most recent commit I renamed `load_minimal` into `load`.
You can also check out `load_required` which will handle the upcasting for you. For instance
```rust
egl::DynamicInstance::<egl::EGL1_4>::load_required()
```
will load EGL, check that it provides at least EGL 1.4 and return a dynamic instance for it. Remember that you can still continue to try to upgrade this instance to an EGL 1.5 instance if needed.

I think these names should be clearer. If this is fine with you, maybe I'll finally be able to publish it :smile:

**Comment by niklaskorz** [comment]
@timothee-haudebourg from my side everything looks fine now, https://github.com/gfx-rs/gfx/pull/3591 works as intended

**Comment by timothee-haudebourg** [comment]
I've just published the version 3.0.0, fixing this issue.
